### ISSUE TITLE
Reactivity/State management in the SDK improvement collection

### ISSUE THREAD

**Issue by samuelstroschein** [inlang/monorepo#1772]
## Problem 

We need to find a better suited state (reactive) management solution for the SDK. 

## Context

The SDK currently uses (Solid) Signals for state management/reactivity. We (knew we would) face numerous issues that pimple up now to the extent that we are inclined to take action. 

## Requirements

- [ ] simple(r) way to prevent/coordinate multiple executions
-- #1740, 
-- https://github.com/inlang/monorepo/issues/1769
- [ ] lazy execution/evaluation for lints
-- https://github.com/inlang/monorepo/issues/1413
- [ ] avoid reactive/not reactive breaking change risk
-- #1680

**Comment by KraXen72** [comment]
using solid signals for an sdk is kind of crazy lol. 
surprised it managed to live for this long 😅

**Comment by NiklasBuchfink** [comment]
I think this is more of a cleaning task and making sure everything is set and passed as it should. We didn't even use stores, because it is complex to understand, but this could help eliminate this issue with handling complex data like our linting.

There is room for improvement. I dug more into it in the last days with podcasts and having the Solid docs as good night reading :)

**Comment by samuelstroschein** [comment]
> using solid signals for an sdk is kind of crazy lol.
> surprised it managed to live for this long 😅

The SDK has the odd requirement that close to every call is reactive. 

Getting the name of a project, querying a message, getting project settings, installing a plugin is/are all reactive properties that should bind to a GUI without leading to constant re-renders. Otherwise, the fink editor and parrot will be noticeably slow. 

The perfect match for our use case does not seem to exist. The question is: What solution comes somewhat close?

**Comment by samuelstroschein** [comment]
Some research over the holidays: 

- we need an async runtime/state management/reactive programming solution 
- scala has multiple e.g. https://typelevel.org/cats-effect/ and https://zio.dev/
- JS ecosystem seems empty handed except for RxJS and the brand new Effect (influenced from Scalas ZIO) https://effect.website/

**Comment by samuelstroschein** [comment]
i just got off a call with Datner regarding effect ts , see his comment [here](https://www.reddit.com/r/typescript/comments/16w3iwn/comment/k42tu1h/?utm_source=share&utm_medium=web2x&context=3) 

- effect ts solves most problems but is not fully stable yet, requires full buy-in and is complex :/
- we need observables/signals somewhere to expose changes to UI frameworks 
- RxJS would be the go-to. 
- we should take a look at xstate 
- [after i raised this idea] going with mobx for simple observables and "manually" handling async state might be a good idea

**mobx vs solid signals**

Both are similar by creating observable state i.e. `createSignal()` or `makeObservable()`. Pros I see over solid signals: 

- production-ready and battle tested
- debugging capabilities https://mobx.js.org/analyzing-reactivity.html
- has async "flows" which can be canceled https://mobx.js.org/actions.html#using-flow-instead-of-async--await-

Cons I see: 
- not async focused :/ 
-- no throttling, debouncing
-- might be okay to manually optimize async flows as needed

**Comment by samuelstroschein** [comment]
### New hard requirement: The solution should not "leak reactivity" to consumers. 

This requirement has been forgotten but was a large driving force for Signals. The SDK should work in plain JS without quirks: 

```ts
// this should work
console.log(project.errors)

// requiring a reactivity pattern should not be required!
project.errors.subscribe((e) => console.log(e))

```

This is a strong reason to choose MobX, which relies on proxies to achieve "reactive magic" but looks like and can work like vanilla JS. The benefits of MobX over Signals have been explored here https://github.com/opral/monorepo/issues/1772#issuecomment-1875981167. 

_This requirement got "re-emerged" in https://github.com/opral/monorepo/issues/1680#issuecomment-1877742428_

**Comment by samuelstroschein** [comment]
I stumbled on https://github.com/artalar/reatom which might also be a worth a look

- explicit reactivity (vs mobx implicit)

- mentions async state solutions that effector apparently fails at

- The context object, while verbose, should make testing easy

- Unfortunately abysmal adoption but seems to be developed and used in production for years

cc @janfjohannes because we chatted about effector

**Comment by samuelstroschein** [comment]
Discussions with @martin-lysk that implicit reactivity (MobX & Signals) seem undesirable 

_https://www.loom.com/share/260363100ff34d3f850f8a4913c851b5?highlightComment=28441771&t=93_

<img width="728" alt="CleanShot 2024-01-09 at 23 33 51@2x" src="https://github.com/opral/monorepo/assets/35429197/92bc4712-79bf-49d7-af29-b78d19abd291">

**Comment by samuelstroschein** [comment]
@jldec wrote in a check in https://discord.com/channels/897438559458430986/1084748995349446667/1202188931362193417

> initial thoughts on reactivity - the apis in solid are not helping us, and I'm not convinced that other similar abstractions like [reatom](https://github.com/artalar/reatom) will help us either. My instinct (not fully confirmed) is to try to eliminate the extra abstraction layer and use explicit callbacks where possible. Automatic dependency tracking based on events is not buying us enough to justify the hidden additional complexity.

We started the SDK without reactivity but needed it to build UI layers with the SDK. It seems like you (like @martin-lysk  and @janfjohannes  before you) overlooked this requirement? 

There needs to be a reactivity pattern that propagates changes to consumers in a **correct** manner. The SDK state is not isolated to itself. I emphasize correct computations here. The majority (all?) of the SDK state is async. Parallel async computations that depend on each other are the norm.  Trying to coordinate them manually sounds like we will inevitably build our own state management solution with many bugs along the road. 

I agree that Solid/Signals is not the right solution. We took them as a quick and dirty way to achieve bindings from the SDK state to the @inlang/fink-editor UI layer. Another requirement btw. Whatever solution we choose needs bindings to UI frameworks like reatom, mobx  & co already have https://www.reatom.dev/package/npm-solid-js/. Developing and maintaining such bindings manually = we are building our own state management solution, which seems out of scope.

**Comment by samuelstroschein** [comment]
Here is an example of "reactivity is needed". 

Also re-stating that this discussion and the comment regarding "reactive system is somehow needed to propagate changes to consumers" (https://github.com/opral/monorepo/issues/1772#issuecomment-1919193325) applies 1:1 to @opral/lix.

- inlang/sdk is the consumer (UI layer) of lix as the SDK is to inlang apps
- Manually tracking changes in `repo.*` to trigger SDK side effects will lead to unmaintainable (callback hell) code.

_With reactivity in lix_

- exposing JS standard similar to `window.*` object (https://github.com/opral/monorepo/issues/2161)
- automatically deals with reactivity 
- IMPORTANT: lix and inlang can use explicit reactivity internally, but it shouldn't leak to the consumer (https://github.com/opral/monorepo/issues/2161)

```ts
// the conditional must be re-run the moment the current branch changes
if (repo.currentBranch !== "main") {
  await loadSettings()
}

// loadSettings needs to re-execute if the settings file changes
async function loadSettings() {
  const file = await repo.fs.readFile("project.inlang/settings.json")
  // ...
}
```

_Without reactivity in lix_

- manual (and correct!) subscribing is necessary
- need to deal with the disposal of subscriptions (will surely be forgotten!)
- nesting of subscriptions can be indefinitely deep. this will be unmaintainable 

```ts
repo.currentBranch.subscribe((branch) => {
  if (repo.currentBranch !== "main"){
     repo.fs.readFile("project.inlang/settings.json").subscribe((file) => {
       await loadSettings()
      })
  }
})


async function loadSettings() {
  const file = await repo.fs.readFile("project.inlang/settings.json")
  // ...
}
```

**Comment by jldec** [comment]
@samuelstroschein thanks for the additional details regarding lix. Could you also clarify (or point to issues) why the current [lix fs watch api](https://github.com/opral/monorepo/blob/main/lix/source-code/fs/src/NodeishFilesystemApi.ts#L26) does not meet our needs? 

I take it from [this comment](https://github.com/opral/monorepo/blob/main/inlang/source-code/sdk/src/loadProject.ts#L110) that a watcher was intended here.

**Comment by samuelstroschein** [comment]
> @jldec  Could you also clarify (or point to issues) why the current [lix fs watch api](https://github.com/opral/monorepo/blob/main/lix/source-code/fs/src/NodeishFilesystemApi.ts#L26) does not meet our needs?

It does. 

My intent was not to surface a problem in the fs API but to show that manual subscriptions will lead to callback hell and bugs. I updated my comment and changed `repo.fs.readFile.subscribe` to `repo.fs.watch`. 

Did I answer your question?

---

_(Not to be discussed in this issue)_

_We could make the point that `repo.fs.readFile()` should be reactive for the same reasons we want `project.*` and `repo.*` props to be reactive: Avoid callback hell, manual diffing, and reduce the likelihood of bugs. Manually setting up watchers and diffing already led to several bugs. A reactive `repo.fs.readFile()` seems like it avoids code like https://github.com/opral/monorepo/pull/2108/files#r1473518448_

**Comment by jldec** [comment]
From https://github.com/opral/monorepo/issues/1772#issuecomment-1919193325
> It seems like you ... overlooked this requirement?

@samuelstroschein regarding the SDK api for syncing with UI, it's clear that we need something there - definitely not overlooked, in fact I expect we'll keep using solid in our UI tier - just not all the way down to the sdk internals.

No firm opinion yet, but I would probably start a POC using somthing similar to one of these database pub-sub apis. .e.g 
- https://supabase.com/docs/guides/realtime/postgres-changes
- https://github.com/soundxyz/redis-pubsub?tab=readme-ov-file#subscribe-and-publish-to-the-channel

**Comment by samuelstroschein** [comment]
> in fact I expect we'll keep using solid in our UI tier

New projects use https://lit.dev/. We want to phase out solid. Just use web components and be done. Any custom renderer like Svelte, React, Solid, Vue is kinda redundant nowadays. 

>  POC using somthing similar to one of these database pub-sub apis. .e.g

But why? 


1. There is no need for a server. The beauty of inlang is precisely having no server. The state is small enough to be handled client-side which removes the entire server layer and makes automation, extension, and collaboration easier. Read [this](https://www.inkandswitch.com/local-first/). 
2. If you mean you want something similar to pub-sub [but client-side], then this is reactivity. Building our own solution for reactivity seems redundant. RxJS, MobX, Effector, Solid, Reatom, solve the problem of "here is a change" -> "you can listen to that change and trigger side effects". 

Let me know if we should hop on a call. There seems to be un-alignment regarding the architecture of the SDK.

**Comment by jldec** [comment]
Alignment :) 

- The message persistence layer does CRUD and never loses data.
- Lint rules are an extension point with a plugin interface
- Lix provides a file-system API backed (lazily) by a git repo
- All of the above are designed to run in the browser

**Comment by NiklasBuchfink** [comment]
I heard in a podcast today that the TC39 works on signal standardization. I hope this is the proposal they were referring to, but it's a good idea to keep it in mind. If a standardization for reactivity and signals comes up, we'll probably adopt it:
- [Decorators proposal](https://github.com/tc39/proposal-decorators)
- Lit already implemented this in the [3.0 prerelease](https://lit.dev/blog/2023-09-27-lit-3.0-prerelease-2/)

**Comment by samuelstroschein** [comment]
Here is an example test that should (must?) pass https://github.com/samuelstroschein/poc-sdk-reactivity with pseudocode, test cases, and a real implementation. 

cc @janfjohannes @jldec

**Comment by martin-lysk** [comment]
Thanks for the poc and the loom. I thought about this and i think we are on a wrong track here.

While we all agree that we need to find a better simpler way to express the dependencies than we have in the sdk I think we mix up initialisation with a running SDK Project.  

**Background:** 
We try to pack into the sdk is asynchron state. But while async and await have simplified triages of api fetches and error handling a lot easier, the underlying concept can become hard to understand / debug and. grasp. 

**Example 1:**
You can create a deadlock in a single thread env. ??? Yes two promises can await each other end never resolve. While the main loop is not blocked your pretty async function may never return. 

**Example 2:** 
The synchronous part of a async function gets called immediately when you call the function but the result is processed only after the next tick if you await it. 

```
async function foo(name) {
  console.log(name, "start");
  await console.log(name, "middle");
  console.log(name, "end");
}

foo("First");
foo("Second");
```
Whats printed out? What do you think first - what is your second guess? Can you explain it without looking at [spec](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await#control_flow_effects_of_await) again. 

<details>

<summary>console</summary>

First start  
First middle
Second start
Second middle
First end
Second end

</details>

Important things to notice: an async function is not asynchronous if it has no await statement inside of it, the order of the execution can not be known from outside without the knowledge of the implementation details. Even if you control the flow from outside - microqueues can always hop in and say - NO NO NO. 

I guess this reflects my first intuition here: https://github.com/opral/monorepo/issues/1846#issuecomment-1848609181 and i try to be more precise on this.

**I would say CRUD operations in the SDK should be sync.** yes this is just half of the truth. No matter how hard we try - the persistence will always be async*. But these are two things. Changing properties in memory - _sync_ and persisting those changes to the disk - _async_.

We load the data into the project object - we keep settings, messages etc. in memory - and whenever we face a change on those objects/properties we persist the changes to the corresponding files on disc again. When we detect a change (fs.watch) we load the changes from the file again. We synchronize but we don't have a guarantee that memory representation and disk are in sync. 

Take the case you outlined in the loom @samuelstroschein, settings change - a language tag was changed. Nothing prevents a separate process to write to the settings file while we do a computation on the messages. Not even checking the settings file before each operation would help us here - we would again end up with a lock file to synchronize our "async" operation. While those simple tests in the POC run green in your env - on a real env with multiple processes messing around with your settings file - this will become flaky again. 

### Proposal

1. A DAG that describes the dependencies of properties in the project. 
2. An async load function and reactive subscribable properties - similar like the once we are having at the moment. 
3. continue to have synchronous CRUD operations that reflect the changes directly after the execution **but** return a promise that stores the changes to disk
4. We use the dependency DAG to describe and detect conflicting changes on the disk that clash with changes in memory.

Take the current implementation of setSettings in loadProject. This is currently synchronus but it doesn't handle errors during disk writes. Also the write to the disc is not finished before the function returns.

```diff

-- const setSettings = (settings: ProjectSettings): Result<void, ProjectSettingsInvalidError> => {
++ const setSettings = (settings: ProjectSettings): Promise<void> => {
--			try {
				const validatedSettings = parseSettings(settings)
				_setSettings(validatedSettings)

--				writeSettingsToDisk(validatedSettings)
--				return { data: undefined }
++				return writeSettingsToDisk(validatedSettings)

--			} catch (error: unknown) {
--				..
--			}
		}
```

Whats cool about this approac. The function is not marked as async - also the affect takes place directly. 

If you set the Language tag by calling setSettings - the new language tag will become visible directly, all subsequent call will work on the new language tag AND the persistence of the new settings is already scheduled and we return the promise that will take care of this. There is no need to synchronize - since its synchron. You can't see a different value as long as you are on the same stack.

This comes with some nice features:

1.) If save messages to disc is throttled you could batch all changes together using Promise.all 

```
await Promise.all([
messageQuery.update({where: { id: "1" }, data: ... };
messageQuery.update({where: { id: "1" }, data: ... };
messageQuery.update({where: { id: "1" }, data: ... };
])
```

2.) If a change should to hit the disc before we can move on - one can use await that ensures that the returned promise - which stores the changes to disk gets resolved before we move on.
`
await project.setSettings(...)`

3.) An exception that happens on the message obejct in memory will be thrown into the callee function and doesn't need to be catched using .catch().
...

**Comment by janfjohannes** [comment]
@NiklasBuchfink  btw pretty sure the decorators you linked to is not what they meant with signal standardization, would be great to find the spec in question.

**Comment by janfjohannes** [comment]
@martin-lysk  i could not really follow your reasoning to have sync actions.  i would argue that running expensive linting and having potential  huge state in a tab will eventually lead to running in a web worker for web an async api can just be moved without changes but this would not be possible with sync

**Comment by martin-lysk** [comment]
Linting, full text search and other computation heavy operation got to be async for sure. But If it comes to the raw message handling - like CRUD there is one big plus on sync: atomicity. 

If we make CRUD async a simple two operation like this can fail:
```
const messageToUpdate = await project.query.messages.get({ id: "b"});
const messageToUpdate.variants[0].pattern[0].value = 'updated text';

// upsert is async -> we schedule it to the next tick -> an fs watch will not know about the planned update
await project.query.messages.upsert({
  where: { id: "b"},
  data: messageToUpdate
})
```

If we have crud sync and only the persistence done async we can handle those cases in persistence layer where all information is available.

**Comment by samuelstroschein** [comment]
@martin-lysk

> I thought about this and i think we are on a wrong track here.

All you propose is what I have in mind. Not sure where we are wrong here. This discussion is more about the reactive state management solution that we should adopt to help us with: 

- dependency tracking
- atomic computations with async side effects that are cancellable. 
For example, after mutating the runtime state, the write-to file fails. The CRUD operations must now fail and show the user "couldn't save your change. please retry". Docs for [[effect ts](https://effect.website/docs/guides/resource-management)], [[reatom's scheduling](https://www.reatom.dev/package/core/#ctxschedule)], and [[reatom's abort controller](https://www.reatom.dev/package/async/#withabort)]
- automatic resource management like debouncing [[reatom docs](https://www.reatom.dev/package/effects/#concurrent)], caching [[reatom docs](https://www.reatom.dev/package/async/#withcache)] [[effect ts docs](https://effect.website/docs/guides/batching-caching)]

@martin-lysk Reatoms scheduling feature seems to be more or less what you describe, but it takes care of dependency tracking, atomicy, and cancellation. 

Furthermore, I think we can rule out MobX and Effector (not Effect TS) based on the insight that **the main problem seems to revolve around async side effects that need be "bottoms up cancellable"**:

```ts
// PSEUDOCODE
async function updateMessage(args) {

  // three side effects. if one of them fails, the entire computation must fail.
  //
  // the ctx.schedule pseudocode API takes care of handling this.
  return await ctx.schedule(async () => {

    await updateStorage(args) 
    
    updateMessagesMap() // @martin-lysk here is your sync 
    
    //💥 throws. the schedule needs to be canceled "bottoms up" because something is broken
    //    and we don't want the user to suddenly have a broken project
    await executePluginsExporters(args) 
  })
 
}
```

### Example: your proposal makes CRUD operations async [by returning a promise], which aligns with #1846 :D

![CleanShot 2024-02-11 at 20 20 23@2x](https://github.com/opral/monorepo/assets/35429197/078347b8-3d9f-42d8-8be2-b9d86fe5162d)

1. you proposal returns a promise?
2. yes, that's why #1846 exists. side-effects need to be awaited.

#1846 argues that any CRUD (really any property of `project.*` btw) API must return a promise because apps need to await *all* async side-effects like persistence to storage. Otherwise, apps are buggy. A user changes a message, the app does not await async side-effects like persistency and wups the user, and inlang app developer wonders why the message is never persistent. 

The following code is a naive bugfix for https://github.com/opral/monorepo/issues/1968 because CRUD ops do not return an promise: "Please let 4 seconds be enough to await all side effects":

https://github.com/opral/monorepo/blob/34b29ddc73993b41da8541367a5beb21c0f92482/inlang/source-code/cli/src/commands/machine/translate.ts#L133-L135

**In case you really argued to leave CRUD functions sync, the bug above gives me 100% confidence that CRUD ops need to be async and await side-effects to avoid bugs.** Issue #1846 already identified multiple bugs.

**Comment by samuelstroschein** [comment]
We should probably open a google doc/spreadsheet to collect requirements at this point.

**Comment by martin-lysk** [comment]
I guess i understand the confusion here:

I am **not** proposing **not to return a promise** I am proposing make the crud functions **non async** functions that return a promise.

SDK functions would have two phases:

1) changing state
I suggest that changes in the sdk should be synchronous. Namely a call like setLanguageTag('en') should change the language tag instantly/synchronous and the effact a call has should always be reflected in the state of the sdk in the callstack we are in and in the tick we are in.
It should even throw synchrousely if you use the function in the wrong way (like setLanguageTag('bla')).

2) persisting state
After the state has been updated retrun a promise that takes care of the persistence 
```js
class Project {

   static async loadProject(inlangFolder: string, lix: Lix) {
      // load settings ...
      const settings = await loadSettings();
      // load messages ...
      const messages = await loadMessages(settings);
      // setup (synchronous) reactivity on properties using a DAG
      // ... the frame work that we choose at the end

      // connect external effects
      lix.watch(inlangFolder, () => {
           // trigger effect based on the changes
      })

      return new Project(settings, messages)
   }
 
inlangFolder: string,
  lix: Lix,
  settings: Settings,

  messages: Messages,
  
   updateRefLanguageTag(languageTag: string) {
      setting.refLanguage = languageTag;
      
       return (async () => {
             await lix.writeFile(inlangFolder + '/settings.json');
       })()
   }
 
   updateMessage(id: string, data: any) {
       if (id.split('_').length != 4) {
         throw ne Error("Id must contain 3 undercores"); 
       }
       messages[id] = data;

      // we can and should use reatom or another library to schedule the save and allow it to be canceled
      // this should only point to the fact that we don't use an async function here and the that chages that the messages is asking to do have hit the sdk before we write the change
       return (async () => {
             await lix.writeFile(inlangFolder + '/message/v1' + messages[id];
       })()
       
   }
}
```

Usage
```
```

**Comment by NiklasBuchfink** [comment]
> @NiklasBuchfink btw pretty sure the decorators you linked to is not what they meant with signal standardization, would be great to find the spec in question.

@janfjohannes You are right, I asked for resources, this was the answer:
https://twitter.com/AtilaFassina/status/1757048374140796952

The video gives a loose indication of where the signals are heading. There has been a [proposal draft for observables](https://github.com/tc39/proposal-observable) in the past, but for signals nothing concrete yet. It's still WIP.

**Comment by samuelstroschein** [comment]
**Sentiment from my after the call earlier today:** 

Reatom seems to address @martin-lysk's raised issues of concurrency and batching natively. 

Take a look at the [concurrent API](https://www.reatom.dev/package/effects/#concurrent) and [async package](https://www.reatom.dev/package/async/#withabort). Both APIs seem to specifically address Martin's concerns about handling concurrent updates. In contrast to Martin's sync proposal, reatom's API comes with explicit handling of cases such as ["last-in-win"](https://www.reatom.dev/package/async/#withabort), or [rollback](https://www.reatom.dev/package/core/#ctxschedule-rollback-api). Reatom's APIs are a bit convoluted/the docs not crisp, but it seems like it solves our issues better than other solutions: 

- seems to address our problems of concurrency, async dependencies, and exposing reactivity (adapters) to clients
- functional programming preferred (in contrast to XState or MobX)
- extensive debugging capabilities via context
- easy migration: seems like we can simply replace solid signals with reatom's "atoms"


**Notes**

- i don't quite get what martin's proposal solves that reatom, rxjs & co don't solve natively
- how a CRUD operation updates state internally (sync, async) is irrelevant to app developers as long as the function operation is yielding correct results and is transparent to the developer that async side-effects need to be awaited.
- the problems @martin-lysk seems to address with his proposal of making something sync or not seem to be addressed by [reatom's schedule API](https://www.reatom.dev/package/core/#ctxschedule)
- an inlang and lix app both need APIs like `await getMessage()` to throw, and if it throws, revert the changes. 


cc @martin-lysk @janfjohannes @jldec

**Comment by samuelstroschein** [comment]
@opral/lix this requirement and example for reactive state came up in the github action that @NiklasBuchfink is building today:

- the project should not be required to be re-opened because the UX in an app like @opral/inlang-fink can become miserable if the project load takes 20 seconds while a reactive branch switch could only take 3 seconds (fictive numbers) 

```ts
const repo = await openRepository({ branch: "main" })
const project await loadProject(repo)

// expect: lint reports on main branch
const reports1 = project.query.messageLintReports.getAll()

await repo.changeBranch("fluffy-gummybear")

// expect: lint reports on fluffy gummybear branch
const reports2 = project.query.messageLintReports.getAll() 

```

**Comment by janfjohannes** [comment]
> @opral/lix this requirement and example for reactive state came up in the github action that @NiklasBuchfink is building today:
> 
> The project should not be required to be re-opened because the UX in an app like @opral/inlang-fink can become miserable if the project load takes 20 seconds while a reactive branch switch could only take 3 seconds (fictive numbers)
> 
> ```ts
> const repo = await openRepository({ branch: "main" })
> const project await loadProject(repo)
> 
> // expect: lint reports on main branch
> const reports1 = project.query.messageLintReports.getAll()
> 
> await repo.changeBranch("fluffy-gummybear")
> 
> // expect: lint reports on fluffy gummybear branch
> const reports2 = project.query.messageLintReports.getAll() 
> ```

@samuelstroschein  of course! this behaviour is allready part of my POC and totally on the same place. fixing   the current requirement to reopen is tracked in this bug : https://github.com/orgs/opral/projects/45/views/1?pane=issue&itemId=53434606&sortedBy%5Bdirection%5D=desc&sortedBy%5BcolumnId%5D=Status

**Comment by artalar** [comment]
Hi! Reatom author there. I would like to help you with this discussion, if you don't mind :)

## Concurrency, rollbacks, async management

Based on what I understand from this discussion, here is a list of points you might find interesting:

- The rollback API (`ctx.schedule(callback, -1)`) only works during the **sync**(!) transaction, which starts with the first action call (or atom update) in the call stack. But.
- We have [CauseContext](https://github.com/artalar/reatom/blob/bcb3d091aea93d25c6dca005e9a1925f25b996ea/packages/effects/src/index.ts#L23), which is kinda ponyfil to [AsyncContext](https://github.com/tc39/proposal-async-context), and you could use it to store all the relative process data. For example, you could use it to store a list of rollback callbacks and automatically call them on all `AsyncAction.onReject`, for example.
- The [async package] (https://www.reatom.dev/package/async/) has been developed and used in some very large and complex applications, so it is very battle-tested and I am proud of its features and quality. One of the most incredible feature is [reatomResource](https://www.reatom.dev/package/async/#reatomresource), which allows you to describe dependent async data short and safe, but with ability to insert custom logic easily, as we just use native `await` ([the debounce example](https://www.reatom.dev/package/effects/#concurrent) shows it perfectly).

## About ecosystem preservation and stability

To be clear, I am really tired of the fragmentation and stability of the React ecosystem and am trying to build something opposite with Reatom. We want to cover all common frontend tasks with a set of well-coupled primitives. I expect that a normal Reatom user will not have any headaches about which library to choose for one or another task, because we will have almost all needed features. And the important thing is that we try to build all packages with enterprise-grade quality. In the last half year, for one site builder I worked on, I personally developed a new forms package, a router package, a zod package (reactive model generations from a zod contract with automatic changes synchronization), and it will be released soon. But there are a lot of contributions with new features from other contributors too! We are [very welcome](https://www.reatom.dev/contributing) to increase our ecosystem and we want to store it all in a stable monorepo. And you should know that you can always write your own alternative solution as the primitives (atoms, ctx, hooks) are public and it is powerful enough.

Currently we have a paid maintainer for some kind of tasks, but we plan to move to https://polar.sh to open doors for other involved people.

In the next few years we plan to grow and grow. Reatom is the core tool of many long-lived projects that it's users and I care about. But it will always be a project with a heart, the heart that cares about web stability, performance and quality.

## About the future

I can't tell you enough to ask all you possible questions, but here is the meaning: in this year we will move from explicit `ctx` **leaving all current features** and bring new incredible once. One of the most important new feature is algebraic effect, like React Suspens. It will be optional feature of course.

---

If you have any questions, please, ping me :)

**Comment by samuelstroschein** [comment]
@artalar Thank you for the reply! 

We figured that we need multi-threading (via workers) to fulfill our state/computation management requirements. Instead of hyperoptimizing the main thread, we will likely go for:

1. Workers compute stuff
2. Bridge between workers and main thread
3. Main thread exposes some reactivity bindings

I am unsure how reatom would help us with 1 and 2. Maybe for 3 it is a good choice. That will be decided by the person evaluating an architecture for the state management solution we come up with.

**Comment by artalar** [comment]
@samuelstroschein, for example, you can use [withBroadcastChannel](https://www.reatom.dev/package/persist-web-storage/#withbroadcastchannel) to synchronize data from a worker to consumers. For posting data processing, you can use actions that will synchronize via `postMessage` (we could develop an adapter for that). The main idea and benefit are to write the logic once but run it in different contexts (worker/tab).
